### PR TITLE
List all causes in verbose errors

### PR DESCRIPTION
**Pull request by charlespierce**
Info
-----
* The verbose error output (and error logs) currently only shows the first level underlying cause of a given error. If the error is more deeply nested, with multiple wrapper `Error`s, then only the first will be listed.

Changes
-----
* Updated the `compose_error_details` method to traverse the entire tree of `.cause()` calls, outputting all of the causes.
* Also refactored the code a bit so `compose_error_details` returns an `Option`, instead of unwrapping the cause outside.

### CODE REVIEW

File: crates/notion-core/src/error/reporter.rs [hunk 1]
```diff
@@ -98,8 +95,31 @@ impl ErrorReporter {
...
+        None => {
+            return None;
+        }
+    };
+    let mut details = String::new();
+
+    // Walk up the tree of causes and include all of them
+    loop {
+        write!(details, "{}", format_error_cause(current))
+            .expect("write! to a String doesn't fail");
```
Comment (mikrostew) [inline crates/notion-core/src/error/reporter.rs]: You can avoid the `write!` and `expect()` here by using string concatenation:

```rust
details += &format_error_cause(current);
```

Or you could use `push_str`

```rust
details.push_str(&format_error_cause(current));
```

File: crates/notion-core/src/error/reporter.rs [hunk 2]
```diff
@@ -98,8 +95,31 @@ impl ErrorReporter {
...
+
+        match current.cause() {
+            Some(cause) => {
+                write!(details, "\n\n").expect("write! to a String doesn't fail");
```
Comment (mikrostew) [inline crates/notion-core/src/error/reporter.rs]: Same thing here

File: crates/notion-core/src/error/reporter.rs (hunk 1 above)
Comment (charlespierce) [reply crates/notion-core/src/error/reporter.rs]: Good call! I was so caught up thinking about formatting that I completely missed that I didn't need the power of the `format!` macros when appending a string value. Updated!

File: crates/notion-core/src/error/reporter.rs (hunk 2 above)
Comment (charlespierce) [reply crates/notion-core/src/error/reporter.rs]: Updated

**Pull request merged**